packages/truffle-db/src/loaders/artifacts/index.ts

```diff
@@ -0,0 +1,255 @@
+  private config: object;
+
+  constructor (db: TruffleDB, config?: object) {
+    this.db = db;
+    this.config = config;
+  }
+
+  async load (): Promise<void> {
+    const {
+      data: {
+        artifacts: {
+          contractNames
+        }
+      }
+    } = await this.db.query(GetContractNames);
+
+    if(this.config) {
+      await this.loadCompilations(this.config);
+    }
+    await this.loadBytecodes(contractNames);
+    await this.loadSources(contractNames);
+
+  }
+
+  async loadBytecodes(contractNames: string[]) {
+    const createBytecodes = await Promise.all(contractNames.map(
+      async (name) =>
+        (await this.db.query(GetBytecode, { name }))
+          .data
+          .artifacts
+          .contract
+          .constructor
+          .createBytecode
+    ));
+
+    const bytecodes = [...createBytecodes];
+
+    await this.db.query(AddBytecodes, { bytecodes });
+  }
+
+  async loadSources(contractNames: string[]) {
+    const contractSources = await Promise.all(contractNames.map(
+      async (name) =>
+        (await this.db.query(GetSource, { name }))
+          .data
+          .artifacts
+          .contract
+          .sourceContract
+          .source
+    ));
+
+    const sources = [...contractSources];
+
+    await this.db.query(AddSources, { sources });
+  }
+  
+  async loadCompilations(compilationConfig: object) {
+    let compilationsArray = [];
+    let sources = [];
+    await Contracts.compile(compilationConfig, async (err, output) => {
+      if(err) console.error(err);
+      else {
+        const compilationsData = Object.values(output.contracts);
+        for(let contract in compilationsData) {
+          //add source from compilation into workspace first, otherwise compilation addition will fail
+          let sourceObject = {
+            contents: compilationsData[contract]["source"], 
+            sourcePath: compilationsData[contract]["sourcePath"]
+          };
+
+          let sourceId = generateId({
+            contents: compilationsData[contract]["source"], 
+            sourcePath: compilationsData[contract]["sourcePath"]
+          });
+          
+          sources.push({contents: compilationsData[contract]["source"], sourcePath: compilationsData[contract]["sourcePath"]});
+
+          let compilationObject = {
+            compiler: {
+              name: compilationsData[contract]["compiler"]["name"],
+              version: compilationsData[contract]["compiler"]["version"]
+            },
+            contracts: [
+              {
+                name: compilationsData[contract]["contract_name"],
+                source: {
+                  id: sourceId
+                },
+                ast: {
+                  json: JSON.stringify(compilationsData[contract]["ast"]),
+                }
+              }
+            ],
+            sources: [
+              { id: sourceId }
+            ]
+          }
+
+          compilationsArray.push(compilationObject);
```

Comment:
I think this isn't what we want... we should have one single compilation with multiple sources, since we only invoke truffle-workflow-compile the once.

Reply:
I initially thought the same but workspace was expecting an array of compilations and I remembered you saying that we should write that code to accept plurals - but of course that is different than this concrete implementation of adding a compilation. Will fix this, thanks!

As a side note, it looks like compiler info is returned from truffle-workflow-compile as part of each contract. I imagine all contracts should have the same compiler, right? So would it be alright to get this info via the first contract in that array? Or perhaps you have something else in mind?

Reply:
In this case, it's safe to rely on the compiler info from the first contract, since it's all part of the same compilation!